File: ghost-engine-kernel/instructions for antigravity/instructions11/instructions11/templates/ge-fetch.js
#!/usr/bin/env node
/**
 * ge-fetch <jobId>
 * Downloads levelSpec from Ghost Engine and saves to GE Doom incoming folder.
 *
 * Env:
 *   API_BASE_URL  (e.g., https://your-alb)
 *   GE_DOOM_DIR   (path to GE Doom project/build root)
 */
import fs from "node:fs";
import path from "node:path";

const base = process.env.API_BASE_URL;
const geDir = process.env.GE_DOOM_DIR || ".";

const jobId = process.argv[2];
if (!base) throw new Error("API_BASE_URL is required");
if (!jobId) throw new Error("Usage: ge-fetch <jobId>");

async function fetchJson(url) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
  return res.json();
}

async function fetchText(url) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
  return res.text();
}

async function main() {
  const status = await fetchJson(`${base}/status/${jobId}`);
  if (status.status !== "done") {
    throw new Error(`Job not done: ${status.status}`);
  }
  const r = status.result || {};
  const incoming = path.join(geDir, "Levels", "incoming");
  fs.mkdirSync(incoming, { recursive: true });

  let levelSpecJson;
  if (r.levelSpecUrl) {
    levelSpecJson = await fetchText(r.levelSpecUrl);
  } else if (r.levelSpec) {
    levelSpecJson = JSON.stringify(r.levelSpec, null, 2);
  } else {
    throw new Error("No levelSpecUrl or inline levelSpec found in result");
  }

  const outPath = path.join(incoming, `${jobId}.levelSpec.json`);
  fs.writeFileSync(outPath, levelSpecJson, "utf-8");
  console.log(`Saved: ${outPath}`);

  // TODO: launch GE Doom with --levelSpec outPath
  // Example:
  //   spawnSync(path.join(geDir, "bin", "GE-Doom"), ["--levelSpec", outPath], { stdio: "inherit" });
  console.log("Next: launch GE Doom with this file (wire your engine's launch command here).");
}

main().catch((e) => {
  console.error("ge-fetch failed:", e);
  process.exit(1);
});

File: ghost-engine-kernel/instructions for antigravity/instructions11/instructions11/templates/ge-fetch.py
#!/usr/bin/env python3
"""ge-fetch <jobId>
Downloads levelSpec from Ghost Engine and saves to GE Doom incoming folder.

Env:
  API_BASE_URL  (e.g., https://your-alb)
  GE_DOOM_DIR   (path to GE Doom project/build root)
"""
import json, os, sys, pathlib
import urllib.request

def http_json(url: str):
    with urllib.request.urlopen(url) as r:
        if r.status != 200:
            raise RuntimeError(f"HTTP {r.status} for {url}")
        return json.loads(r.read().decode("utf-8"))

def http_text(url: str) -> str:
    with urllib.request.urlopen(url) as r:
        if r.status != 200:
            raise RuntimeError(f"HTTP {r.status} for {url}")
        return r.read().decode("utf-8")

def main():
    base = os.environ.get("API_BASE_URL")
    ge_dir = os.environ.get("GE_DOOM_DIR", ".")
    if not base:
        raise RuntimeError("API_BASE_URL is required")
    if len(sys.argv) < 2:
        raise RuntimeError("Usage: ge-fetch <jobId>")
    job_id = sys.argv[1]

    status = http_json(f"{base}/status/{job_id}")
    if status.get("status") != "done":
        raise RuntimeError(f"Job not done: {status.get('status')}")

    r = status.get("result", {})
    incoming = pathlib.Path(ge_dir) / "Levels" / "incoming"
    incoming.mkdir(parents=True, exist_ok=True)

    if "levelSpecUrl" in r and r["levelSpecUrl"]:
        level_spec = http_text(r["levelSpecUrl"])
    elif "levelSpec" in r and r["levelSpec"]:
        level_spec = json.dumps(r["levelSpec"], indent=2)
    else:
        raise RuntimeError("No levelSpecUrl or inline levelSpec")

    out_path = incoming / f"{job_id}.levelSpec.json"
    out_path.write_text(level_spec, encoding="utf-8")
    print(f"Saved: {out_path}")

    # TODO: launch GE Doom with --levelSpec out_path
    print("Next: launch GE Doom with this file (wire your engine's launch command here).")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print("ge-fetch failed:", e, file=sys.stderr)
        sys.exit(1)
